File: backend/analytics/simulator.py
"""What-if simulator and the build/buy calculator.

Both are *models*, not measurements, and every response says so. The what-if
simulator's defaults are seeded from recorded traffic where it exists, so the
starting point is the client's own shape rather than a vendor's.
"""
import math
from dataclasses import dataclass

from backend.analytics.aggregate import _f, live
from backend.optimizer import pricing


def defaults_from(records: list, registry) -> dict:
    """Seed the simulator from real traffic when there is any."""
    rs = live(records)
    served = [r for r in rs if not r.get("cache_hit")]
    premium = registry.tier_default("premium")
    if not served:
        return {"requests_per_month": 1_000_000, "input_tokens": 2000, "output_tokens": 500,
                "cache_hit_rate": 20.0, "context_reduction_pct": 25.0, "batch_pct": 15.0,
                "mix": {"cheap": 60.0, "medium": 30.0, "premium": 10.0},
                "baseline_model": premium.id, "seeded_from": "none (illustrative defaults)"}
    n = len(rs)
    mix = {"cheap": 0.0, "medium": 0.0, "premium": 0.0}
    for r in served:
        t = r.get("tier")
        if t in mix:
            mix[t] += 1
    tot = sum(mix.values()) or 1
    inp = sum(int(_f(r, "input_tokens", 0)) for r in served) / max(1, len(served))
    base_inp = sum(int(_f(r, "baseline_input_tokens", 0)) for r in served) / max(1, len(served))
    out = sum(int(_f(r, "output_tokens", 0)) for r in served) / max(1, len(served))
    return {
        "requests_per_month": 1_000_000,
        "input_tokens": round(base_inp or inp),
        "output_tokens": round(out) or 250,
        "cache_hit_rate": round(100 * sum(1 for r in rs if r.get("cache_hit")) / n, 1),
        "context_reduction_pct": round(100 * (base_inp - inp) / base_inp, 1) if base_inp else 0.0,
        "batch_pct": round(100 * sum(1 for r in rs if r.get("execution_mode") in ("batch", "offline")) / n, 1),
        "mix": {k: round(100 * v / tot, 1) for k, v in mix.items()},
        "baseline_model": premium.id,
        "seeded_from": f"{n} recorded request(s)",
    }


def simulate(params: dict, registry) -> dict:
    """Model monthly spend for a baseline and an optimised configuration.

    The arithmetic is deliberately transparent and shown in the response so a
    client can check it: no hidden multipliers.
    """
    reqs = max(0, int(params.get("requests_per_month", 1_000_000)))
    inp = max(0, int(params.get("input_tokens", 2000)))
    out = max(0, int(params.get("output_tokens", 500)))
    cache = min(100.0, max(0.0, float(params.get("cache_hit_rate", 20.0)))) / 100
    ctx = min(95.0, max(0.0, float(params.get("context_reduction_pct", 25.0)))) / 100
    batch = min(100.0, max(0.0, float(params.get("batch_pct", 15.0)))) / 100
    mix = params.get("mix") or {"cheap": 60, "medium": 30, "premium": 10}
    total_mix = sum(max(0.0, float(v)) for v in mix.values()) or 1.0
    mix = {k: max(0.0, float(v)) / total_mix for k, v in mix.items()}
    base_model = registry.get(params.get("baseline_model") or registry.tier_default("premium").id)

    baseline_per_req = pricing.price(base_model, inp, out, registry=registry).total_usd
    baseline = reqs * baseline_per_req

    served = reqs * (1 - cache)
    inp_opt = round(inp * (1 - ctx))
    lines, optimised = [], 0.0
    for tier, share in mix.items():
        try:
            m = registry.tier_default(tier)
        except StopIteration:
            continue
        n_tier = served * share
        n_batch = n_tier * batch
        n_sync = n_tier - n_batch
        c_sync = pricing.price(m, inp_opt, out, registry=registry).total_usd
        c_batch = pricing.price(m, inp_opt, out, execution_mode="batch", registry=registry).total_usd
        cost = n_sync * c_sync + n_batch * c_batch
        optimised += cost
        lines.append({"tier": tier, "model": m.id, "share_pct": round(100 * share, 1),
                      "requests": round(n_tier), "batch_requests": round(n_batch),
                      "cost_per_request_usd": round(c_sync, 8),
                      "batch_cost_per_request_usd": round(c_batch, 8),
                      "monthly_usd": round(cost, 2),
                      "batch_discount": registry.provider(m.provider).batch_discount})
    saving = baseline - optimised
    # Contribution of each lever, computed by turning it off one at a time.
    def _spend(cache_v=cache, ctx_v=ctx, batch_v=batch, mix_v=mix):
        s = reqs * (1 - cache_v)
        i = round(inp * (1 - ctx_v))
        tot = 0.0
        for tier, share in mix_v.items():
            try:
                m = registry.tier_default(tier)
            except StopIteration:
                continue
            nt = s * share
            nb = nt * batch_v
            tot += (nt - nb) * pricing.price(m, i, out, registry=registry).total_usd
            tot += nb * pricing.price(m, i, out, execution_mode="batch", registry=registry).total_usd
        return tot
    all_premium = {"cheap": 0.0, "medium": 0.0, "premium": 1.0}
    levers = [
        {"lever": "Cache", "usd": round(_spend(cache_v=0.0) - optimised, 2)},
        {"lever": "Context reduction", "usd": round(_spend(ctx_v=0.0) - optimised, 2)},
        {"lever": "Model routing", "usd": round(_spend(mix_v=all_premium) - optimised, 2)},
        {"lever": "Batch execution", "usd": round(_spend(batch_v=0.0) - optimised, 2)},
    ]
    return {
        "basis": "modeled",
        "note": "A model of monthly spend under the stated assumptions, not a measurement. Prices come "
                "from the versioned registry; real savings depend on the traffic mix and quality "
                "requirements of the actual workload.",
        "price_registry_version": registry.version,
        "inputs": {"requests_per_month": reqs, "input_tokens": inp, "output_tokens": out,
                   "cache_hit_rate_pct": round(100 * cache, 1),
                   "context_reduction_pct": round(100 * ctx, 1), "batch_pct": round(100 * batch, 1),
                   "mix_pct": {k: round(100 * v, 1) for k, v in mix.items()},
                   "baseline_model": base_model.id},
        "baseline": {"model": base_model.id, "cost_per_request_usd": round(baseline_per_req, 8),
                     "monthly_usd": round(baseline, 2), "annual_usd": round(baseline * 12, 2)},
        "optimized": {"monthly_usd": round(optimised, 2), "annual_usd": round(optimised * 12, 2),
                      "cost_per_request_usd": round(optimised / reqs, 8) if reqs else 0.0,
                      "cached_requests": round(reqs * cache), "served_requests": round(served),
                      "input_tokens_per_request": inp_opt, "lines": lines},
        "savings": {"monthly_usd": round(saving, 2), "annual_usd": round(saving * 12, 2),
                    "pct": round(100 * saving / baseline, 1) if baseline else 0.0},
        "lever_contributions": sorted(levers, key=lambda l: -l["usd"]),
    }


def build_vs_buy(params: dict, registry) -> dict:
    """Fully loaded self-hosting cost against the equivalent API spend.

    Self-hosting is only cheaper past a real utilisation threshold, and the
    calculator is built to show that rather than to argue for it.
    """
    reqs_day = max(0.0, float(params.get("requests_per_day", 100_000)))
    inp = max(0, int(params.get("input_tokens", 2000)))
    out = max(0, int(params.get("output_tokens", 500)))
    gpu_hourly = max(0.0, float(params.get("gpu_hourly_usd", 2.99)))
    gpus = max(1, int(params.get("gpu_count", 1)))
    util = min(100.0, max(1.0, float(params.get("utilization_pct", 40.0)))) / 100
    throughput = max(1.0, float(params.get("tokens_per_second_per_gpu", 2500)))
    eng_monthly = max(0.0, float(params.get("engineering_monthly_usd", 12_000)))
    ops_multiplier = max(1.0, float(params.get("ops_multiplier", 1.0)))
    storage = max(0.0, float(params.get("storage_monthly_usd", 200)))
    egress = max(0.0, float(params.get("egress_monthly_usd", 150)))
    observability = max(0.0, float(params.get("observability_monthly_usd", 300)))
    redundancy = max(1.0, float(params.get("redundancy_factor", 1.5)))
    api_model_id = params.get("api_model") or registry.tier_default("medium").id
    api_model = registry.get(api_model_id)

    tokens_day = reqs_day * (inp + out)
    monthly_reqs = reqs_day * 30.4
    api_monthly = monthly_reqs * pricing.price(api_model, inp, out, registry=registry).total_usd

    gpu_monthly = gpu_hourly * 24 * 30.4 * gpus * redundancy
    infra = gpu_monthly + storage + egress + observability
    people = eng_monthly * ops_multiplier
    self_monthly = infra + people
    capacity_tokens_day = throughput * 86400 * gpus * util
    capacity_ok = capacity_tokens_day >= tokens_day
    needed_gpus = math.ceil(tokens_day / (throughput * 86400 * util)) if throughput and util else None
    per_million_self = (self_monthly / (tokens_day * 30.4 / 1e6)) if tokens_day else None
    per_million_api = pricing.price(api_model, 1_000_000, 0, registry=registry).total_usd + \
        pricing.price(api_model, 0, 1_000_000, registry=registry).total_usd * (out / max(1, inp + out))

    # Break-even: daily tokens at which self-hosting matches the API bill.
    api_per_token = (api_monthly / (tokens_day * 30.4)) if tokens_day else 0.0
    breakeven_tokens_day = (self_monthly / 30.4 / api_per_token) if api_per_token else None
    breakeven_requests_day = (breakeven_tokens_day / (inp + out)) if breakeven_tokens_day and (inp + out) else None

    verdict = "api"
    if capacity_ok and self_monthly < api_monthly:
        verdict = "self_host"
    elif not capacity_ok:
        verdict = "api_capacity"
    return {
        "basis": "modeled",
        "note": "Fully loaded comparison. Self-hosting is only cheaper past a real utilisation "
                "threshold; engineering and on-call time usually dominate the GPU bill at POC volume.",
        "inputs": {**params, "api_model": api_model.id},
        "api": {"model": api_model.id, "monthly_usd": round(api_monthly, 2),
                "annual_usd": round(api_monthly * 12, 2),
                "per_million_tokens_usd": round(per_million_api, 4)},
        "self_hosted": {
            "monthly_usd": round(self_monthly, 2), "annual_usd": round(self_monthly * 12, 2),
            "gpu_monthly_usd": round(gpu_monthly, 2), "people_monthly_usd": round(people, 2),
            "other_monthly_usd": round(storage + egress + observability, 2),
            "people_share_pct": round(100 * people / self_monthly, 1) if self_monthly else 0.0,
            "per_million_tokens_usd": round(per_million_self, 4) if per_million_self else None,
            "capacity_tokens_per_day": round(capacity_tokens_day),
            "required_tokens_per_day": round(tokens_day),
            "capacity_sufficient": capacity_ok,
            "gpus_required": needed_gpus,
        },
        "breakeven": {"tokens_per_day": round(breakeven_tokens_day) if breakeven_tokens_day else None,
                      "requests_per_day": round(breakeven_requests_day) if breakeven_requests_day else None,
                      "current_tokens_per_day": round(tokens_day)},
        "verdict": verdict,
        "verdict_text": {
            "api": f"Stay on the API. Self-hosting would cost ${self_monthly:,.0f}/month against "
                   f"${api_monthly:,.0f}/month, and {round(100 * people / self_monthly) if self_monthly else 0}% "
                   f"of that is people, not GPUs.",
            "api_capacity": f"Stay on the API. The configured fleet serves "
                            f"{round(capacity_tokens_day):,} tokens/day at {round(100 * util)}% utilisation "
                            f"but the workload needs {round(tokens_day):,}; "
                            f"{needed_gpus} GPU(s) would be required.",
            "self_host": f"Self-hosting is cheaper at this volume: ${self_monthly:,.0f}/month against "
                         f"${api_monthly:,.0f}/month. Confirm the utilisation assumption "
                         f"({round(100 * util)}%) holds at the traffic's real peak-to-trough shape.",
        }[verdict],
    }

File: frontend/app/pages/simulator.js
import { get, post } from '../api.js';
import { card, chip, esc, field, int, money, pct, slider, stat, table } from '../ui.js';
import { ranked } from '../charts.js';

let params = null;
let mode = 'whatif';

export async function render(root, state) {
  mode = state.params.mode || 'whatif';
  if (!params) {
    try { params = await get('/api/simulate/defaults'); } catch { params = {}; }
  }
  root.innerHTML = `
  <div class="page-head">
    <div><div class="kicker">Simulator · modelled, not measured</div>
      <h1>What would this cost at your volume?</h1>
      <p class="lede">Everything on this page is a model of spend under assumptions you control, not
        a measurement of your traffic. The defaults are seeded from the requests this system has
        actually recorded, so the starting point is real even though the projection is not.</p></div>
    <div class="seg">
      <button class="${mode === 'whatif' ? 'on' : ''}" data-go="simulator"
        data-params='{"mode":"whatif"}'>What-if</button>
      <button class="${mode === 'buildbuy' ? 'on' : ''}" data-go="simulator"
        data-params='{"mode":"buildbuy"}'>Build vs buy</button>
    </div>
  </div>
  <div id="body"></div>`;
  if (mode === 'whatif') await whatif(root);
  else await buildBuy(root);
}

// ------------------------------------------------------------------ what-if
async function whatif(root) {
  const body = root.querySelector('#body');
  const p = params;
  body.innerHTML = `<div class="grid g-1-2">
    <div>${card('Assumptions', `
      <div class="note" style="margin-bottom:14px">Seeded from ${esc(p.seeded_from || 'defaults')}.</div>
      ${field('Requests per month', `<input type="number" id="reqs" value="${p.requests_per_month}" step="10000">`)}
      <div class="grid g2" style="gap:10px;margin-top:10px">
        ${field('Input tokens', `<input type="number" id="inp" value="${p.input_tokens}" step="100">`)}
        ${field('Output tokens', `<input type="number" id="out" value="${p.output_tokens}" step="50">`)}
      </div>
      <div style="margin-top:14px">
        ${sliderRow('cache', 'Cache hit rate', p.cache_hit_rate)}
        ${sliderRow('ctx', 'Context reduction', p.context_reduction_pct)}
        ${sliderRow('batch', 'Batch-eligible traffic', p.batch_pct)}
      </div>
      <div class="kicker" style="margin:16px 0 8px">Model mix of served traffic</div>
      ${sliderRow('cheap', 'Cheap tier', p.mix.cheap)}
      ${sliderRow('medium', 'Balanced tier', p.mix.medium)}
      ${sliderRow('premium', 'Frontier tier', p.mix.premium)}
      <div class="note" style="margin-top:10px">The mix is normalised, so it never has to add to 100.</div>
    `)}</div>
    <div id="out-panel">${card('', '<div class="empty"><span class="spin"></span></div>', { pad: false })}</div>
  </div>`;

  const inputs = [...body.querySelectorAll('input')];
  let t;
  const run = async () => {
    const payload = {
      requests_per_month: num('#reqs'), input_tokens: num('#inp'), output_tokens: num('#out'),
      cache_hit_rate: num('#cache'), context_reduction_pct: num('#ctx'), batch_pct: num('#batch'),
      mix: { cheap: num('#cheap'), medium: num('#medium'), premium: num('#premium') },
      baseline_model: p.baseline_model,
    };
    body.querySelectorAll('[data-out]').forEach((el) => {
      el.textContent = `${document.querySelector(`#${el.dataset.out}`).value}%`;
    });
    try {
      const r = await post('/api/simulate', payload);
      body.querySelector('#out-panel').innerHTML = whatifResult(r);
    } catch (e) {
      body.querySelector('#out-panel').innerHTML = `<div class="callout risk">${esc(e.message)}</div>`;
    }
  };
  const num = (sel) => Number(body.querySelector(sel).value) || 0;
  inputs.forEach((i) => {
    i.oninput = () => { clearTimeout(t); t = setTimeout(run, 90); };
  });
  await run();
}

const sliderRow = (id, label, value) => `
  <label class="field" style="margin-bottom:10px">
    <span class="lab">${esc(label)}<b data-out="${id}">${value}%</b></span>
    ${slider(id, value, { min: 0, max: 100, step: 1 })}</label>`;

function whatifResult(r) {
  return card('Modelled monthly spend', `
    <div class="grid g3" style="gap:10px">
      ${stat({ label: 'Baseline', value: money(r.baseline.monthly_usd, { big: true }),
        sub: `${esc(r.baseline.model)} for everything` })}
      ${stat({ label: 'Optimized', value: money(r.optimized.monthly_usd, { big: true }), tone: 'good',
        sub: `${int(r.optimized.cached_requests)} cached, ${int(r.optimized.served_requests)} served` })}
      ${stat({ label: 'Saving', value: pct(r.savings.pct), hero: true,
        sub: `${money(r.savings.monthly_usd, { big: true })}/month · ${money(r.savings.annual_usd, { big: true })}/year` })}
    </div>
    <div class="section" style="margin-top:20px">
      <h2 style="font-size:13px">What each lever contributes</h2>
      <p class="note" style="margin:2px 0 12px">Computed by turning each lever off on its own, so
        overlapping effects are not double-counted.</p>
      ${ranked(r.lever_contributions.map((l) => ({ label: l.lever, value: l.usd })),
        { valueFmt: (v) => money(v, { big: true }) })}
    </div>
    <div class="section" style="margin-top:20px">
      <h2 style="font-size:13px">Per tier</h2>
      ${table([
        { label: 'Tier', cell: (l) => `${chip(l.tier)} <span class="note">${esc(l.model)}</span>` },
        { label: 'Share', align: 'r', cell: (l) => `<span class="num">${pct(l.share_pct, 0)}</span>` },
        { label: 'Requests', align: 'r', cell: (l) => `<span class="num">${int(l.requests)}</span>` },
        { label: 'Per request', align: 'r', cell: (l) => `<span class="num">${money(l.cost_per_request_usd)}</span>` },
        { label: 'Batch', align: 'r', cell: (l) => l.batch_discount
            ? `<span class="num" style="color:var(--save)">−${Math.round(l.batch_discount * 100)}%</span>`
            : '<span class="note">none</span>' },
        { label: 'Monthly', align: 'r', cell: (l) => `<b class="num">${money(l.monthly_usd, { big: true })}</b>` },
      ], r.optimized.lines, { minWidth: 560 })}
    </div>
    <div class="callout warn" style="margin-top:16px"><b>Read this as a sensitivity scenario, not a
      promise.</b> ${esc(r.note)}</div>`, { sub: `prices ${esc(r.price_registry_version)}` });
}

// --------------------------------------------------------------- build/buy
async function buildBuy(root) {
  const body = root.querySelector('#body');
  const defaults = { requests_per_day: 100000, input_tokens: 2000, output_tokens: 500,
    gpu_hourly_usd: 2.99, gpu_count: 2, utilization_pct: 40, tokens_per_second_per_gpu: 2500,
    engineering_monthly_usd: 12000, redundancy_factor: 1.5 };
  body.innerHTML = `<div class="grid g-1-2">
    <div>${card('Assumptions', `
      ${field('Requests per day', `<input type="number" id="rpd" value="${defaults.requests_per_day}" step="10000">`)}
      <div class="grid g2" style="gap:10px;margin-top:10px">
        ${field('Input tokens', `<input type="number" id="bi" value="${defaults.input_tokens}" step="100">`)}
        ${field('Output tokens', `<input type="number" id="bo" value="${defaults.output_tokens}" step="50">`)}
      </div>
      <div class="grid g2" style="gap:10px;margin-top:10px">
        ${field('GPU $/hour', `<input type="number" id="gh" value="${defaults.gpu_hourly_usd}" step="0.1">`)}
        ${field('GPU count', `<input type="number" id="gc" value="${defaults.gpu_count}" step="1" min="1">`)}
      </div>
      <div style="margin-top:12px">${sliderRow('util', 'Utilisation', defaults.utilization_pct)}</div>
      <div class="grid g2" style="gap:10px">
        ${field('Tokens/s per GPU', `<input type="number" id="tps" value="${defaults.tokens_per_second_per_gpu}" step="100">`)}
        ${field('Engineering $/month', `<input type="number" id="eng" value="${defaults.engineering_monthly_usd}" step="1000">`)}
      </div>
      <div class="note" style="margin-top:12px">Engineering and on-call are the line item that
        decides this, and the one most build/buy cases leave out.</div>`)}</div>
    <div id="bb-out">${card('', '<div class="empty"><span class="spin"></span></div>', { pad: false })}</div>
  </div>`;

  let t;
  const run = async () => {
    body.querySelectorAll('[data-out]').forEach((el) => {
      el.textContent = `${document.querySelector(`#${el.dataset.out}`).value}%`;
    });
    const n = (s) => Number(body.querySelector(s).value) || 0;
    try {
      const r = await post('/api/simulate/build-vs-buy', {
        requests_per_day: n('#rpd'), input_tokens: n('#bi'), output_tokens: n('#bo'),
        gpu_hourly_usd: n('#gh'), gpu_count: n('#gc'), utilization_pct: n('#util'),
        tokens_per_second_per_gpu: n('#tps'), engineering_monthly_usd: n('#eng'),
      });
      body.querySelector('#bb-out').innerHTML = bbResult(r);
    } catch (e) {
      body.querySelector('#bb-out').innerHTML = `<div class="callout risk">${esc(e.message)}</div>`;
    }
  };
  body.querySelectorAll('input').forEach((i) => {
    i.oninput = () => { clearTimeout(t); t = setTimeout(run, 90); };
  });
  await run();
}

function bbResult(r) {
  const selfCheaper = r.verdict === 'self_host';
  return card('API against self-hosting', `
    <div class="callout ${selfCheaper ? 'save' : 'warn'}" style="margin-bottom:16px">
      <b>${selfCheaper ? 'Self-hosting wins at this volume.' : 'Stay on the API.'}</b>
      ${esc(r.verdict_text)}
    </div>
    <div class="grid g2" style="gap:10px">
      ${stat({ label: 'API', value: money(r.api.monthly_usd, { big: true }),
        sub: `${esc(r.api.model)} · ${money(r.api.annual_usd, { big: true })}/year` })}
      ${stat({ label: 'Self-hosted, fully loaded', value: money(r.self_hosted.monthly_usd, { big: true }),
        tone: selfCheaper ? 'good' : 'warn',
        sub: `${pct(r.self_hosted.people_share_pct, 0)} of it is people, not GPUs` })}
    </div>
    <div class="section" style="margin-top:18px">
      <h2 style="font-size:13px">Where the self-hosted money goes</h2>
      ${ranked([
        { label: 'GPUs (with redundancy)', value: r.self_hosted.gpu_monthly_usd, color: 'var(--spend)' },
        { label: 'Engineering and on-call', value: r.self_hosted.people_monthly_usd, color: 'var(--risk)' },
        { label: 'Storage, egress, observability', value: r.self_hosted.other_monthly_usd, color: 'var(--warn)' },
      ], { valueFmt: (v) => money(v, { big: true }) })}
    </div>
    <div class="section" style="margin-top:18px">
      <h2 style="font-size:13px">Capacity and break-even</h2>
      ${table([
        { label: '', cell: (x) => `<b>${esc(x[0])}</b>` },
        { label: '', align: 'r', cell: (x) => `<span class="num">${x[1]}</span>` },
      ], [
        ['Workload', `${int(r.self_hosted.required_tokens_per_day)} tokens/day`],
        ['Fleet capacity at this utilisation', `${int(r.self_hosted.capacity_tokens_per_day)} tokens/day`],
        ['GPUs actually required', `${int(r.self_hosted.gpus_required)}`],
        ['Break-even volume', r.breakeven.requests_per_day
          ? `${int(r.breakeven.requests_per_day)} requests/day` : 'n/a'],
        ['Current volume', `${int(r.breakeven.current_tokens_per_day / (r.inputs.input_tokens + r.inputs.output_tokens))} requests/day`],
      ], { minWidth: 400 })}
    </div>
    <div class="note" style="margin-top:14px">${esc(r.note)}</div>`);
}
